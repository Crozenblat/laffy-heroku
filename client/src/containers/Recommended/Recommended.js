import React, {useState, useEffect} from "react";
import styled from "styled-components";
import axios from "axios";

import withErrorHandler from "../../hoc/withErrorHandler";

import InfoMessage from "../../components/Messages/InfoMessage";
import ComedianWindow from "../../components/Comedian/ComedianWindow";
import LoadingIcon from "../../components/LoadingIcon/LoadingIcon";

const Rccmmndd = styled.div`
    display: flex;
    flex-wrap: wrap;
    width: 100%;

    @media only screen and (max-width: 426px){
        width: 85%;
        justify-content: space-around;
    };
`;

const Recommended = props => {
    const [content, setContent] = useState(<LoadingIcon/>);

    useEffect(() => {
        (async () => {
            try{
                let recommended = await axios.get("/user/recommended", {withCredentials: true});
                if(recommended.data.length){
                    setContent(recommended.data.map((comedian, index) => {
                        return <ComedianWindow key={`recommended-${index}`} name={comedian.name} accountImage={comedian.accountImage} tags={comedian.tags} linkId={comedian._id}/>
                    }));
                }else{
                    setContent(<InfoMessage>You don't currently have any recommendations. Go favorite some comedians to get recommendations!</InfoMessage>);
                };
            }catch(err){
                setContent(<InfoMessage>Recommended comedians could not be fetched. Please refresh the page.</InfoMessage>);
                console.log(err);
            };
        })();
    }, []);

    return (
        <Rccmmndd>
            {content}
        </Rccmmndd>
    )
}

export default withErrorHandler(Recommended);