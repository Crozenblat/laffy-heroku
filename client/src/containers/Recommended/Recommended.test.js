import React from "react";
import {render, screen} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import {BrowserRouter} from "react-router-dom";

import seed from "../../../backend/seed";

import Recommended from "./Recommended";

describe("Recommended Comedians Container", () => {
    const userSeventeen = seed.users[16];
    
    it("should render correctly", () => {
        let wrapper = render(
            <BrowserRouter>
                <Recommended userId={userSeventeen._id}/>
            </BrowserRouter>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it("should display the top 5 recommended comedian windows for a logged in user", async () => {
        render(
            <BrowserRouter>
                <Recommended userId={userSeventeen._id}/>
            </BrowserRouter>
        );
        let recommendedComedians = await screen.findAllByTestId(/comedian-window/);

        expect(recommendedComedians).toHaveLength(5);

        test.each([
            [recommendedComedians[0], "Marc Maron", seed.comedians[11].tags[0]],
            [recommendedComedians[1], "Patton Oswalt", seed.comedians[12].tags[0]],
            [recommendedComedians[2], "D.L. Hughley", seed.comedians[13].tags[0]],
            [recommendedComedians[3], "John Mulaney", seed.comedians[14].tags[0]],
            [recommendedComedians[4], "Katherine Ryan", seed.comedians[15].tags[0]],
        ])("recently viewed comedian %s is displayed", (recentlyViewed, comedian, tag) => {
            expect(recentlyViewed).toContainElement(screen.getByAltText(comedian));
            expect(recentlyViewed).toContainElement(screen.getByText(comedian));
            expect(recentlyViewed).toContainElement(screen.getByRole("button", {name: tag}));
        });
    });

    it("should display a message to user who hasn't favorited any comedians encouraging them to favorite some comedians to recieve recommendations", async () => {
        render(
            <BrowserRouter>
                <Recommended userId={null}/>
            </BrowserRouter>
        );
        let recommendedComedians = await screen.findAllByTestId(/comedian-window/);
        let message = screen.getByText("You haven't favorited any comedians yet. Go favorite some so we can recommended more that we thnk you'll like!");
        let exploreBtn = screen.getByRole("button", {name: "Explore"});

        expect(recommendedComedians).toHaveLength(0);
        expect(message).toBeInTheDocument();
        expect(exploreBtn).toBeInTheDocument();
    });

    it("should display a message to a non-logged in user encouraging them to sign up for an account", () => {
        render(
            <BrowserRouter>
                <Recommended userId={null}/>
            </BrowserRouter>
        );
        let message = screen.getByText("Whoops! It looks like you're not currently logged in. Sign up now!");
        let signUpBtn = screen.getByRole("button", {name: "Sign Up!"});

        expect(message).toBeInTheDocument();
        expect(signUpBtn).toBeInTheDocument();
    });
});