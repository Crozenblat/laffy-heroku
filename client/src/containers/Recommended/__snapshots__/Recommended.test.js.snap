// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommended Comedians Container should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-AxheI ieTQbu"
      >
        <a
          class="sc-AxgMl fQrWMw"
          href="/comedians?name=undefined"
        >
          <div
            class="sc-AxhCb bPffyD"
          >
            <span
              class="sc-AxjAm bIFvHb"
              type="comedianWindow"
            />
            <h3
              class="sc-AxhUy bEaupT"
            >
              Comedian
            </h3>
            <div>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Crude
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Smart
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Boston
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Italian
                </button>
              </a>
            </div>
          </div>
        </a>
        <a
          class="sc-AxgMl fQrWMw"
          href="/comedians?name=undefined"
        >
          <div
            class="sc-AxhCb bPffyD"
          >
            <span
              class="sc-AxjAm bIFvHb"
              type="comedianWindow"
            />
            <h3
              class="sc-AxhUy bEaupT"
            >
              Comedian
            </h3>
            <div>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Crude
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Smart
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Boston
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Italian
                </button>
              </a>
            </div>
          </div>
        </a>
        <a
          class="sc-AxgMl fQrWMw"
          href="/comedians?name=undefined"
        >
          <div
            class="sc-AxhCb bPffyD"
          >
            <span
              class="sc-AxjAm bIFvHb"
              type="comedianWindow"
            />
            <h3
              class="sc-AxhUy bEaupT"
            >
              Comedian
            </h3>
            <div>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Crude
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Smart
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Boston
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Italian
                </button>
              </a>
            </div>
          </div>
        </a>
        <a
          class="sc-AxgMl fQrWMw"
          href="/comedians?name=undefined"
        >
          <div
            class="sc-AxhCb bPffyD"
          >
            <span
              class="sc-AxjAm bIFvHb"
              type="comedianWindow"
            />
            <h3
              class="sc-AxhUy bEaupT"
            >
              Comedian
            </h3>
            <div>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Crude
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Smart
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Boston
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Italian
                </button>
              </a>
            </div>
          </div>
        </a>
        <a
          class="sc-AxgMl fQrWMw"
          href="/comedians?name=undefined"
        >
          <div
            class="sc-AxhCb bPffyD"
          >
            <span
              class="sc-AxjAm bIFvHb"
              type="comedianWindow"
            />
            <h3
              class="sc-AxhUy bEaupT"
            >
              Comedian
            </h3>
            <div>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Crude
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Smart
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Boston
                </button>
              </a>
              <a
                class="sc-AxirZ iONWjK"
                href=""
              >
                <button
                  class="sc-AxiKw dMhYsm"
                >
                  Italian
                </button>
              </a>
            </div>
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AxheI ieTQbu"
    >
      <a
        class="sc-AxgMl fQrWMw"
        href="/comedians?name=undefined"
      >
        <div
          class="sc-AxhCb bPffyD"
        >
          <span
            class="sc-AxjAm bIFvHb"
            type="comedianWindow"
          />
          <h3
            class="sc-AxhUy bEaupT"
          >
            Comedian
          </h3>
          <div>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Crude
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Smart
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Boston
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Italian
              </button>
            </a>
          </div>
        </div>
      </a>
      <a
        class="sc-AxgMl fQrWMw"
        href="/comedians?name=undefined"
      >
        <div
          class="sc-AxhCb bPffyD"
        >
          <span
            class="sc-AxjAm bIFvHb"
            type="comedianWindow"
          />
          <h3
            class="sc-AxhUy bEaupT"
          >
            Comedian
          </h3>
          <div>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Crude
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Smart
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Boston
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Italian
              </button>
            </a>
          </div>
        </div>
      </a>
      <a
        class="sc-AxgMl fQrWMw"
        href="/comedians?name=undefined"
      >
        <div
          class="sc-AxhCb bPffyD"
        >
          <span
            class="sc-AxjAm bIFvHb"
            type="comedianWindow"
          />
          <h3
            class="sc-AxhUy bEaupT"
          >
            Comedian
          </h3>
          <div>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Crude
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Smart
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Boston
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Italian
              </button>
            </a>
          </div>
        </div>
      </a>
      <a
        class="sc-AxgMl fQrWMw"
        href="/comedians?name=undefined"
      >
        <div
          class="sc-AxhCb bPffyD"
        >
          <span
            class="sc-AxjAm bIFvHb"
            type="comedianWindow"
          />
          <h3
            class="sc-AxhUy bEaupT"
          >
            Comedian
          </h3>
          <div>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Crude
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Smart
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Boston
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Italian
              </button>
            </a>
          </div>
        </div>
      </a>
      <a
        class="sc-AxgMl fQrWMw"
        href="/comedians?name=undefined"
      >
        <div
          class="sc-AxhCb bPffyD"
        >
          <span
            class="sc-AxjAm bIFvHb"
            type="comedianWindow"
          />
          <h3
            class="sc-AxhUy bEaupT"
          >
            Comedian
          </h3>
          <div>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Crude
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Smart
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Boston
              </button>
            </a>
            <a
              class="sc-AxirZ iONWjK"
              href=""
            >
              <button
                class="sc-AxiKw dMhYsm"
              >
                Italian
              </button>
            </a>
          </div>
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
